File: components/page-map/page-map.clj
;; generate a map of a page with links to content
(defn page-map [compiled-html file]
  (let [tags (html/find-tags compiled-html :h1 :h2 :h3 :h4 :h5 :h6)
        path-link (:link file)]
    [:div.sitemap-container
     [:span.sitemap-title "In this article"]
     [:table.sitemap
      (map (fn [tag]
             (let [tag-id (:id (second tag))
                   tag-text (or
                             (and (> 2 (count tag)) (nth tag 2))
                             tag-id)]
               [:tr
                [:td {:id (str "page-map-" tag-id)} [:a {:href (str path-link "#" tag-id)} tag-text]]]))
           tags)]]))

(defn main [file files file-idx compiled-html]
  {:depends-on []
   :body (page-map compiled-html file)})

File: jake-site-deploy/components/PageMap/PageMap.ts
import { findTags } from "html";

const makePageMap = ({ articleHtml, file, rootDir, sourceDir }) => {
  const tags = findTags(articleHtml, ["h1", "h2", "h3", "h4", "h5", "h6"]);
  const link = file.htmlUrl({ rootDir, sourceDir });

  return [
    "div",
    { class: "sitemap-container" },
    ["span", { class: "sitemap-title" }, "In this article"],
    [
      "table",
      { class: "sitemap" }["tbody"].concat(
        tags.map((tag) => [
          "tr",
          ["td", { class: "sitemap-indent" }, tag.depth],
          [
            "td",
            { class: "sitemap-link" },
            ["a", { href: `${link}#${tag.id}` }, tag.text],
          ],
        ])
      ),
    ],
  ];
};

const PageMap = (args) => ({
  dependsOn: [],
  body: makePageMap(args),
});

export default PageMap;
